enterprise/tailnet/pgcoord.go

```diff
@@ -609,62 +803,87 @@ type countedMapper struct {
 	cancel context.CancelFunc
 }
 
-func newQuerier(
-	ctx context.Context, logger slog.Logger,
-	ps pubsub.Pubsub, store database.Store,
-	self uuid.UUID, newConnections chan *connIO, numWorkers int,
-	firstHeartbeat chan<- struct{},
+func newQuerier(ctx context.Context,
+	logger slog.Logger,
+	coordinatorID uuid.UUID,
+	ps pubsub.Pubsub,
+	store database.Store,
+	self uuid.UUID,
+	newConnections chan agpl.Queue,
+	subscriptions chan subscribe,
+	numWorkers int,
+	firstHeartbeat chan struct{},
 ) *querier {
 	updates := make(chan hbUpdate)
 	q := &querier{
-		ctx:            ctx,
-		logger:         logger.Named("querier"),
-		pubsub:         ps,
-		store:          store,
-		newConnections: newConnections,
-		workQ:          newWorkQ[mKey](ctx),
-		heartbeats:     newHeartbeats(ctx, logger, ps, store, self, updates, firstHeartbeat),
-		mappers:        make(map[mKey]*countedMapper),
-		conns:          make(map[*connIO]struct{}),
-		updates:        updates,
-		healthy:        true, // assume we start healthy
+		ctx:                 ctx,
+		logger:              logger.Named("querier"),
+		coordinatorID:       coordinatorID,
+		pubsub:              ps,
+		store:               store,
+		newConnections:      newConnections,
+		subscriptions:       subscriptions,
+		workQ:               newWorkQ[mKey](ctx),
+		heartbeats:          newHeartbeats(ctx, logger, ps, store, self, updates, firstHeartbeat),
+		mappers:             make(map[mKey]*countedMapper),
+		conns:               make(map[uuid.UUID]agpl.Queue),
+		updates:             updates,
+		clientSubscriptions: make(map[uuid.UUID]map[uuid.UUID]struct{}),
+		healthy:             true, // assume we start healthy
 	}
 	q.subscribe()
-	go q.handleConnIO()
-	for i := 0; i < numWorkers; i++ {
-		go q.worker()
-	}
-	go q.handleUpdates()
+
+	go func() {
+		<-firstHeartbeat
+		go q.handleIncoming()
+		for i := 0; i < numWorkers; i++ {
+			go q.worker()
+		}
+		go q.handleUpdates()
+	}()
 	return q
 }
 
-func (q *querier) handleConnIO() {
+func (q *querier) handleIncoming() {
 	for {
 		select {
 		case <-q.ctx.Done():
 			return
+
 		case c := <-q.newConnections:
-			q.newConn(c)
+			switch c.Kind() {
+			case agpl.QueueKindAgent:
+				q.newAgentConn(c)
+			case agpl.QueueKindClient:
+				q.newClientConn(c)
+			default:
+				panic(fmt.Sprint("unreachable: invalid queue kind ", c.Kind()))
+			}
+
+		case sub := <-q.subscriptions:
+			if sub.active {
+				q.newClientSubscription(sub.q, sub.agentID)
+			} else {
+				q.removeClientSubscription(sub.q, sub.agentID)
```

Comment:
It's technically handled by checking the `clientSubscriptions` map, since there will never be an agent subscription with `uuid.Nil`. I had an explicit check but just replaced it with the map lookup, since it essentially does the same thing. Might be good to comment.